Sources/BSON/BSONDocumentIterator.swift

```diff
@@ -28,6 +13,53 @@ public struct BSONDocumentIterator: IteratorProtocol {
         self.buffer.moveReaderIndex(to: 4)
     }
 
+    internal init(over doc: BSONDocument) {
+        self = BSONDocumentIterator(over: doc.buffer)
+    }
+
+    /// Returns the current value's type. Assumes the iterator is in a valid position.
+    internal var currentType: BSONType {
+        guard let typeByte = self.buffer.getInteger(at: self.buffer.readerIndex, as: UInt8.self) else {
+            return .invalid
+        }
+        guard let type = BSONType(rawValue: typeByte) else {
+            return .invalid
+        }
+        return type
+    }
+
+    /// Returns the current key. Assumes the iterator is in a valid position.
+    internal var currentKey: String {
+        guard self.buffer.readableBytes != 1 && !self.exhausted else {
+            return ""
+        }
+        do {
+            return try self.buffer.getCString(at: self.buffer.readerIndex + 1)
+        } catch {
+            fatalError("Cannot read current key \(error.localizedDescription)")
+        }
+    }
+
+    /// Returns the current value. Assumes the iterator is in a valid position.
+    internal var currentValue: BSON {
+        var mutBuffer = self.buffer
+
+        let typeIndicator = self.currentType
+        guard typeIndicator != .invalid else {
+            return .null
+        }
+        do {
+            _ = mutBuffer.readBytes(length: 1) // type indicator
+            _ = try mutBuffer.readCString() // BSON Key
+            guard let type = BSON.allBSONTypes[typeIndicator] else {
+                throw BSONError.InternalError(message: "Unable to read type: \(typeIndicator)")
+            }
+            return try type.read(from: &mutBuffer)
+        } catch {
+            fatalError("Cannot obtain current value: \(error.localizedDescription)")
+        }
+    }
```

Comment:
This presented a bit of an issue, since we only have read and not a "get" the ByteBuffer needs to be mutable in order to read out the current type so I created a copy that's mutable, works fine but I wasn't sure if that's ideal.

Reply:
This property seems to be a relic of having to interact with libbson. If you look at all the existing usages, its always preceded by a call to `advance`. We can make this a single operation (i.e. `next`) to avoid the duplication, and just remove this property altogether I think.